### PR TITLE
Feature/193 other details validation ehrstatus attr

### PR DESCRIPTION
**Pull request by chevalleyc**
## Changes

This fixes multiple issues related to CR #193

Most important change is the new DB layout for table STATUS: added archetype_node_id and name column.

## Related issue

closes: https://github.com/ehrbase/project_management/issues/193

Edit by @jakesmolka:
fixes https://github.com/ehrbase/project_management/issues/157
fixes https://github.com/ehrbase/project_management/issues/158
fixes https://github.com/ehrbase/project_management/issues/159

## Additional information and checks

- [ ] Pull request linked in changelog

### CODE REVIEW

File: api/src/main/java/org/ehrbase/api/service/ValidationService.java [hunk 1]
```diff
@@ -56,6 +57,14 @@
      */
     void check(Composition composition) throws Exception;
 
+    /**
+     * initially check if ehrstatus is valid for further processing
+     * @param ehrStatus
+     * @throws IllegalArgumentException
+     */
+    void check(EhrStatus ehrStatus);
```
Comment (jakesmolka) [inline api/src/main/java/org/ehrbase/api/service/ValidationService.java:65]: I removed the `throws Exception` here since it only throws `RuntimeException`s anyway. Alright? Would that make sense for the `check(Composition composition)`, too?

File: service/src/main/java/org/ehrbase/dao/access/jooq/StatusAccess.java [hunk 2]
```diff
@@ -120,12 +121,21 @@ public static I_StatusAccess retrieveInstanceByParty(I_DomainAccess domainAccess
...
+            if (record == null)
+                return null;
 
-        return createStatusAccessForRetrieval(domainAccess, record);
+            return createStatusAccessForRetrieval(domainAccess, record);
+        }
+        catch (Exception e){
+            if (e instanceof SQLException){
+                //retrieve the cause and use the cause
+                throw new IllegalArgumentException(e.getCause().getMessage());
```
Comment (jakesmolka) [inline service/src/main/java/org/ehrbase/dao/access/jooq/StatusAccess.java:135]: Please use our own exceptions.

File: service/src/main/java/org/ehrbase/service/EhrServiceImp.java [hunk 3]
```diff
@@ -58,22 +57,35 @@
...
+            validationService.check(status);
+        } catch (Exception e) {
+            // rethrow if this class, but wrap all others in InternalServerException
+            if (e.getClass().equals(UnprocessableEntityException.class))
+                throw (UnprocessableEntityException) e;
+            if (e.getClass().equals(IllegalArgumentException.class))
+                throw new ValidationException(e);
+            if (e.getClass().equals(ValidationException.class))
+                throw e;
+            else if (e.getClass().equals(org.ehrbase.validation.constraints.wrappers.ValidationException.class))
```
Comment (jakesmolka) [inline service/src/main/java/org/ehrbase/service/EhrServiceImp.java:83]: We have two kinds of `ValidationException`, which is confusing. Do you know if we could replace the module specific `org.ehrbase.validation.constraints.wrappers.ValidationException` one with our project-wide `org.ehrbase.api.exception.ValidationException` one? I mean not just here, but generally?

File: service/src/main/java/org/ehrbase/service/ValidationServiceImp.java [hunk 4]
```diff
@@ -131,6 +135,44 @@ public void check(Composition composition) throws Exception {
...
+        //first, check the built EhrStatus using the general Archie RM-Validator
+        List<RMObjectValidationMessage> rmObjectValidationMessages = RM_OBJECT_VALIDATOR.validate(ehrStatus);
+
+        if (!rmObjectValidationMessages.isEmpty()){
+            StringBuilder stringBuilder = new StringBuilder();
+            for (RMObjectValidationMessage rmObjectValidationMessage: rmObjectValidationMessages){
+                stringBuilder.append(rmObjectValidationMessage.toString());
+                stringBuilder.append("\n");
+            }
+            throw new ValidationException(stringBuilder.toString());
```
Comment (jakesmolka) [inline service/src/main/java/org/ehrbase/service/ValidationServiceImp.java:154]: I changed the exceptions to our own one. This is our general approach we decided on once. Would you agree here, or had you some custom use case in mind? If not, would you change the exception type in the other methods here, too?

File: service/src/main/java/org/ehrbase/service/ValidationServiceImp.java [hunk 5]
```diff
@@ -131,6 +135,44 @@ public void check(Composition composition) throws Exception {
...
     }
 
+    @Override
+    public void check(EhrStatus ehrStatus) {
+
+        //case of a system generated ehr
+        if (ehrStatus == null)
+            return;
+
+        //first, check the built EhrStatus using the general Archie RM-Validator
```
Comment (jakesmolka) [inline service/src/main/java/org/ehrbase/service/ValidationServiceImp.java:145]: I changed the order of the checks. Now, first, the general RM validation is done and secondly, our more custom rules are applied. Alright?
The other way around I found it hard to differentiate between what I need to keep in mind and check with our `if`s VS. what the archie validator is going to check anyway. Now the broader and general check is applied first and probably catches all usual problems.

File: service/src/main/java/org/ehrbase/service/ValidationServiceImp.java [hunk 6]
```diff
@@ -131,6 +135,44 @@ public void check(Composition composition) throws Exception {
...
+        if (ehrStatus.getSubject().getExternalRef() != null) {  // external_ref has 0..1 multiplicity, so null itself is okay
+            // but if it is there it has to have an ID
+            if (ehrStatus.getSubject().getExternalRef().getId() == null || ehrStatus.getSubject().getExternalRef().getId().getValue().isEmpty())
+                throw new ValidationException("ExternalRef ID is required");
+            // and a namespace
+            if (ehrStatus.getSubject().getExternalRef().getNamespace() == null) {
+                throw new ValidationException("ExternalRef namespace is required");
+                // which needs to be valid
+            } else if (!NAMESPACE_PATTERN.matcher(ehrStatus.getSubject().getExternalRef().getNamespace()).matches())
+                throw new ValidationException("Subject's namespace format invalid");
```
Comment (jakesmolka) [inline service/src/main/java/org/ehrbase/service/ValidationServiceImp.java:171]: I added the namespace check here, which was needed by an integration test. I reordered the conditionals here a bit.

File: tests/robot/EHR_SERVICE_TESTS/B.1_CREATE_EHR/B.1__a)_New_EHR.robot [hunk 7]
```diff
@@ -342,7 +326,7 @@ MF-025 - Create new EHR w/ invalid subject (POST /ehr variants)
     [Documentation]     Covers edge case where subject is provided but is invalid
     ...                 because of some mandatory elements missing.
     [Tags]
-    [Template]          create ehr from data table
+    [Template]          create ehr from data table (invalid)
```
Comment (jakesmolka) [inline tests/robot/EHR_SERVICE_TESTS/B.1_CREATE_EHR/B.1__a)_New_EHR.robot:329]: @testautomation needs to do a quick sanity check on the changes of the test itself

File: service/src/main/java/org/ehrbase/service/ValidationServiceImp.java [hunk 8]
```diff
@@ -131,6 +135,44 @@ public void check(Composition composition) throws Exception {
...
+            return;
+
+        //first, check the built EhrStatus using the general Archie RM-Validator
+        List<RMObjectValidationMessage> rmObjectValidationMessages = RM_OBJECT_VALIDATOR.validate(ehrStatus);
+
+        if (!rmObjectValidationMessages.isEmpty()){
+            StringBuilder stringBuilder = new StringBuilder();
+            for (RMObjectValidationMessage rmObjectValidationMessage: rmObjectValidationMessages){
+                stringBuilder.append(rmObjectValidationMessage.toString());
+                stringBuilder.append("\n");
```
Comment (jakesmolka) [inline service/src/main/java/org/ehrbase/service/ValidationServiceImp.java:152]: I observed a wrong input body (f.e. without `name`) returns the message from this string builder like intended, but the new lines are just literal `\n`, instead of a line break.

**Comment by chevalleyc** [comment]
@jakesmolka thanks for the changes! I will merge this PR as to prevent it to block the next one (in particular PR #209 )
